packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -0,0 +1,438 @@
+import { parseVatSlot } from './parseVatSlots.js';
+
+/*
+  Theory of Operation (imports/Presences)
+
+    Imports are in one of 5 states: UNKNOWN, REACHABLE, UNREACHABLE,
+    COLLECTED, FINALIZED. Note that there's no actual state machine with those
+    values, and we can't observe all of the transitions from JavaScript, but
+    we can describe what operations could cause a transition, and what our
+    observations allow us to deduce about the state:
+
+    * UNKNOWN moves to REACHABLE when a crank introduces a new import
+    * userspace holds a reference only in REACHABLE
+    * REACHABLE moves to UNREACHABLE only during a userspace crank
+    * UNREACHABLE moves to COLLECTED when GC runs, which queues the finalizer
+    * COLLECTED moves to FINALIZED when a new turn runs the finalizer
+    * liveslots moves from FINALIZED to UNKNOWN by syscalling dropImports
+
+    convertSlotToVal either imports a vref for the first time, or
+    re-introduces a previously-seen vref. It transitions from:
+
+    * UNKNOWN to REACHABLE by creating a new Presence
+    * UNREACHABLE to REACHABLE by re-using the old Presence that userspace
+      forgot about
+    * COLLECTED/FINALIZED to REACHABLE by creating a new Presence
+
+    Our tracking tables hold data that depends on the current state:
+
+    * slotToVal holds a WeakRef in [REACHABLE, UNREACHABLE, COLLECTED]
+    * that WeakRef .deref()s into something in [REACHABLE, UNREACHABLE]
+    * possiblyDeadSet holds the vref only in FINALIZED
+    * (TODO) re-introduction could remove the vref from possiblyDeadSet
+
+    Each state thus has a set of perhaps-measurable properties:
+
+    * UNKNOWN: slotToVal[baseRef] is missing, baseRef not in possiblyDeadSet
+    * REACHABLE: slotToVal has live weakref, userspace can reach
+    * UNREACHABLE: slotToVal has live weakref, userspace cannot reach
+    * COLLECTED: slotToVal[baseRef] has dead weakref
+    * FINALIZED: slotToVal[baseRef] is missing, baseRef is in possiblyDeadSet
```

Comment:
```suggestion
    * UNKNOWN: slotToVal[baseRef] is missing, baseRef not in possiblyDeadSet
    * REACHABLE: slotToVal[baseRef] weakref is live, userspace can reach
    * UNREACHABLE: slotToVal[baseRef] weakref is live, userspace cannot reach
    * COLLECTED: slotToVal[baseRef] weakref is dead
    * FINALIZED: slotToVal[baseRef] is missing, baseRef is in possiblyDeadSet
```
Also, I observe a switch here to "baseRef" (from "vref" in preceding text). Can we consistently adopt a single term that covers both virtual/durable cohorts and import/export values? Is "baseRef" already that term? (_"For Presence- and Remotable- style objects, the vref and baseref are the same" below suggests yes_)

Reply:
Hm, good question, I don't have a satisfying answer. Yeah, `baseRef` is the superset, `vref` is one subtype, and we don't have a name for the other subtype.

| category | typical BaseRef | typical related vrefs |
| --- | --- | --- |
| Presence | o-1 | o-1 |
| Remotable / heap | o+1 | o+1 |
| Representative | o+d4/5 | o+d4/5:0 , o+d4/5:1 , o+d4/5:2 |

`slotToVal` is keyed by things in the second column, `valToSlot` has values from the third column. If you call `ceaseRecognition()` with something from the second column, it will act as if you'd called it multiple times with each of the things from the third column.

I think I'll file an issue to refactor our documentation and find a way to be clear about our types. It might take the form of phrases like "the one vref of this baseref (one, because the baseref is Presence- or Remotable- style)".